File: destack-ts/src/encoder/proto/encoder.ts
import { getObjectKey, PROTO_OBJECT_ENCODERS } from "@destack/encoder/proto/generate";
import { loadEncoders } from "@destack/encoder/proto/generated";
import {
  BuiltinObject,
  Encoder,
  NodeType,
  ObjectKind,
  Session,
  StructType,
  Type,
} from "@destack/language";
import { AnyNodeProto, AnyStructProto } from "@destack/proto";

/** Encoder for our protobuf format. */
export class ProtoEncoder implements Encoder<any> {
  constructor() {
    loadEncoders();
  }

  packObject(
    kind: ObjectKind,
    metatype: NodeType | StructType,
    object: BuiltinObject,
  ): AnyStructProto | AnyNodeProto {
    const encoder = PROTO_OBJECT_ENCODERS[getObjectKey(kind, metatype)];
    if (!encoder) {
      throw new Error(
        `no ProtoEncoder for ${ObjectKind[kind] ?? kind}:${NodeType[metatype] ?? StructType[metatype] ?? metatype}`,
      );
    }
    return encoder.packObject(object);
  }

  packObjectBinary(
    kind: ObjectKind,
    metatype: NodeType | StructType,
    object: BuiltinObject,
  ): Uint8Array {
    const encoder = PROTO_OBJECT_ENCODERS[getObjectKey(kind, metatype)];
    if (!encoder) {
      throw new Error(
        `no ProtoEncoder for ${ObjectKind[kind] ?? kind}:${NodeType[metatype] ?? StructType[metatype] ?? metatype}`,
      );
    }
    return encoder.packObjectBinary(object);
  }

  unpackObject(
    kind: ObjectKind,
    metatype: NodeType | StructType,
    value: AnyStructProto | AnyNodeProto,
    session: Session | null,
  ): BuiltinObject {
    const encoder = PROTO_OBJECT_ENCODERS[getObjectKey(kind, metatype)];
    if (!encoder) {
      throw new Error(
        `no ProtoEncoder for ${ObjectKind[kind] ?? kind}:${NodeType[metatype] ?? StructType[metatype] ?? metatype}`,
      );
    }
    return encoder.unpackObject(value, session);
  }

  unpackObjectBinary(
    kind: ObjectKind,
    metatype: NodeType | StructType,
    value: Uint8Array,
    session: Session | null,
  ): BuiltinObject {
    const encoder = PROTO_OBJECT_ENCODERS[getObjectKey(kind, metatype)];
    if (!encoder) {
      throw new Error(
        `no ProtoEncoder for ${ObjectKind[kind] ?? kind}:${NodeType[metatype] ?? StructType[metatype] ?? metatype}`,
      );
    }
    return encoder.unpackObjectBinary(value, session);
  }

  packValue(value: any, type: Type): any {
    throw new Error("not implemented");
  }

  packValueBytes(value: any, type: Type): Uint8Array {
    throw new Error("not implemented");
  }

  unpackValue(type: Type, value: any): any {
    throw new Error("not implemented");
  }

  unpackValueBytes(type: Type, value: Uint8Array): any {
    throw new Error("not implemented");
  }
}

File: destack-py/destack/encoder/proto/encoder.py
from typing import Any, ClassVar, override

from destack.language.core import (
    BuiltinObject,
    Encoder,
    Encoding,
    Graph,
    GraphConnection,
    NodeType,
    ObjectKind,
    Session,
    StructType,
    Type,
)
from destack.proto import AnyObjectProto

from ..cson import CsonEncoder
from .generate import PROTO_CLASSES, PROTO_OBJECT_ENCODERS

_cson_encoder = CsonEncoder()  # for generic Values


class ProtoEncoder(Encoder[AnyObjectProto]):
    """Encoder for our protobuf format."""

    encoding: ClassVar[Encoding] = Encoding.PROTO

    @override
    def pack_object(
        self,
        kind: ObjectKind,
        metatype: NodeType | StructType,
        object: BuiltinObject,
    ) -> AnyObjectProto:
        encoder = PROTO_OBJECT_ENCODERS[kind, metatype]
        return encoder.pack_object(object)

    @override
    def pack_object_bytes(
        self,
        kind: ObjectKind,
        metatype: NodeType | StructType,
        object: BuiltinObject,
    ) -> bytes:
        encoder = PROTO_OBJECT_ENCODERS[kind, metatype]
        object_packed = encoder.pack_object(object)
        return object_packed.SerializeToString()

    @override
    def unpack_object(
        self,
        kind: ObjectKind,
        metatype: NodeType | StructType,
        value: AnyObjectProto,
        *,
        session: Session | None,
        graph: Graph | None,
        connection: GraphConnection | None,
    ) -> BuiltinObject:
        encoder = PROTO_OBJECT_ENCODERS[kind, metatype]
        return encoder.unpack_object(value, session, graph, connection)

    @override
    def unpack_object_bytes(
        self,
        kind: ObjectKind,
        metatype: NodeType | StructType,
        value: bytes,
        *,
        session: Session | None,
        graph: Graph | None,
        connection: GraphConnection | None,
    ) -> BuiltinObject:
        encoder = PROTO_OBJECT_ENCODERS[kind, metatype]
        proto_cls = PROTO_CLASSES[kind, metatype]
        value_decoded = proto_cls.FromString(value)
        return encoder.unpack_object(value_decoded, session, graph, connection)

    @override
    def pack_value(
        self,
        value: Any,
        type: Type,
    ) -> Any:
        raise NotImplementedError

    @override
    def pack_value_bytes(
        self,
        value: Any,
        type: Type,
    ) -> bytes:
        raise NotImplementedError

    @override
    def unpack_value(
        self,
        type: Type,
        value: Any,
        *,
        session: Session | None,
        graph: Graph | None,
        connection: GraphConnection | None,
    ) -> Any:
        raise NotImplementedError

    @override
    def unpack_value_bytes(
        self,
        type: Type,
        value: bytes,
        *,
        session: Session | None,
        graph: Graph | None,
        connection: GraphConnection | None,
    ) -> Any:
        raise NotImplementedError
